### PR TITLE
New Analysis Class API

### PR DESCRIPTION
**Pull request by RHammond2**
This PR addresses a longstanding issue with a lack of consistency between the analysis classes, and the differences between what is passed through `__init__()` vs `run()`. The updated paradigm is as follows:
- All run and __init__ arguments are now able to be passed to the __init__ method
- Only parameters that do not change the nature of what data is validated/checked for existence can be passed to `run()`. For example, the keyword argument `wind_direction_col` can be passed during initialization of `WakeLosses`, but not during the `run` method because we are changing the underlying data.
- Changing reanalysis products is still allowed in `run()` because this is about the sensitivity to long term weather data, and not pertaining to the underlying wind power plant data.
- Just as in v2, analysis classes do not operate on the `PlantData` object, but use the underlying data to create the permutations necessary for the focal analysis.

What this means is that run() can be used to iterate over permutations of an analysis on the same data, but if we need to swap out any data, e.g. different wind direction columns, additional SCADA columns, & etc., that should be considered to be an entirely new analysis, and therefore a new analysis class object is required.

Updated features:
- `openoa/schema/metadata.py:ANALYSIS_REQUIREMENTS` contains all the available modifiers for running analyses. For example, there is now "WakeLosses-scada" and "WakeLosses-tower" to indicate which data will be used, depending on if your wind direction is coming from the `scada` or `tower`.
- `analysis_class.plant` is now a deep copy of the passed data object to ensure that no underlying data are updated. The only exception to `analysis.plant` remaining unchanged is in the case of `PlantData.analysis_type` gets updated to include the analysis that will be performed.

Additionally, this PR expands the analysis requirements settings per #241.

### CODE REVIEW

**Review by ejsimley — CHANGES_REQUESTED** [review]
Hi Rob, I like the consistent format for initializing the analysis classes and specifying/changing parameters in the `run` methods! I have some minor questions and comments, but I think this is mostly ready to go. I will also update the doc strings for the wake losses class and/or change the default values for some of the arguments since the defaults without UQ weren't originally always the average of the default range with UQ.

File: openoa/analysis/aep.py [hunk 1]
```diff
@@ -1003,7 +1098,10 @@ def run_AEP_monte_carlo(self):
             iav[n] = gross_lt_annual.std() / gross_lt_annual.mean()
             avail_pct[n] = avail_lt_losses
             curt_pct[n] = curt_lt_losses
-            lt_por_ratio[n] = (gross_lt.sum() / self._run.num_years_windiness) / gross_por.sum()
+            gross_por_sum = gross_por.sum()
+            if isinstance(gross_por_sum, pd.Series):
```
[the remark this answers is not in the document]
Comment (RHammond2) [reply openoa/analysis/aep.py:1102]: In the future pandas will just break for the original usage when `gross_por_sum` is a single element Series.

File: openoa/analysis/electrical_losses.py [hunk 2]
```diff
@@ -72,16 +72,16 @@ class ElectricalLosses(FromDictMixin):
     """
 
     plant: PlantData
-    UQ: bool = field(default=False, converter=bool)
+    UQ: bool = field(default=False, validator=attrs.validators.instance_of(bool))
     num_sim: int = field(default=20000, converter=int)
-    uncertainty_correction_threshold: NDArrayFloat | tuple[float, float] | float = field(
-        default=0.95, validator=(validate_UQ_input, validate_open_range_0_1)
+    uncertainty_meter: NDArrayFloat | float = field(
+        default=0.005, validator=validate_half_closed_0_1_right
```
[the remark this answers is not in the document]
Comment (RHammond2) [reply openoa/analysis/electrical_losses.py:78]: Thanks for catching this, you're right, this is just missing the update after copying and pasting!

File: openoa/analysis/aep.py [hunk 3]
```diff
@@ -266,23 +276,108 @@ def __attrs_post_init__(self):
...
+        uncertainty_outlier: float | tuple[float, float] = None,
+        uncertainty_nan_energy: float = None,
+        time_resolution: str = None,
+        end_date_lt: str | pd.Timestamp | None = None,
+        ml_setup_kwargs: dict = None,
+    ) -> None:
         """
-        Perform pre-processing of data into an internal representation for which the analysis can run more quickly.
+        Perform pre-processing of data into an internal representation for which the analysis can
+        run more quickly.
```
[the remark this answers is not in the document]
Comment (RHammond2) [reply openoa/analysis/aep.py:314]: This is likely just a remnant of old language that I just skipped over, but I'll update it while we're all here.

File: openoa/analysis/aep.py [hunk 4]
```diff
@@ -266,23 +276,108 @@ def __attrs_post_init__(self):
...
+                " as an input. 'product' is the empty default value and is reserved."
+            )
+
     @logged_method_call
-    def run(self, num_sim: int, reanalysis_subset: list[str] = None):
+    def run(
+        self,
+        num_sim: int,
+        reg_model: str = None,
+        reanalysis_products: list[str] = None,
```
[the remark this answers is not in the document]
Comment (RHammond2) [reply openoa/analysis/aep.py:300]: That's a great point! I think I originally envisioned this being as a subset of any that were originally provided, so if you add more, the method wouldn't work, but this was never actually implemented either.

I'm curious as to your thoughts here on whether it's an appropriate breaking of the paradigm or not. My own thoughts are that reanalysis data is already validated, so it's all there to be used, but at the same time it's still changing the nature of the analysis.

File: openoa/analysis/turbine_long_term_gross_energy.py [hunk 5]
```diff
@@ -159,19 +170,78 @@ def __attrs_post_init__(self):
         logger.info("Processing SCADA data into dictionaries by turbine (this can take a while)")
         self.sort_scada_by_turbine()
 
+    def finalize_reanalysis_products(self):
```
[the remark this answers is not in the document]
Comment (RHammond2) [reply openoa/analysis/turbine_long_term_gross_energy.py:173]: We definitely could, and I think my main hangup is that it's not universally shared across the analysis classes, nor is it doing much to begin with, so I felt it was easier to just keep in the methods where it was needed. That said, I could be convinced if you think we should make another mixin class like with the `FromDictMixin` that manages the dictionary initialization processes.

File: openoa/analysis/_analysis_validators.py [hunk 6]
```diff
@@ -30,13 +30,20 @@ def validate_UQ_input(cls, attribute: attrs.Attribute, value: float | tuple) ->
...
     else:
-        if not isinstance(value, float):
+        # For some defaults values of analysis classes, a tuple is provided, so ensure that those
+        # are correctly converted to mean value with a 2-decimal precision as is intended through
+        # the original API, otherwise raise an error that a float should be provided.
+        if isinstance(value, tuple):
+            if len(value) == 2:
+                object.__setattr__(cls, attribute.name, round(np.mean(value), 2))
+                return
+        if not isinstance(value, (int, float)):
```
[the remark this answers is not in the document]
Comment (RHammond2) [reply openoa/analysis/_analysis_validators.py:40]: Fair point, I think I overthought it when I did that.

File: openoa/analysis/_analysis_validators.py [hunk 7]
```diff
@@ -58,3 +65,27 @@ def validate_open_range_0_1(cls, attribute: attrs.Attribute, value: float | tupl
...
+        attribute (attrs.Attribute): The attrs Attribute information for the class attribute being
+            validated.
+        value (float | tuple): The user input to the class attribute.
+
+    Raises:
+        ValueError: Raised if a single input is passed and outside the range of [0, 1).
+        ValueError: Raised if any of the inputs in the input tuple are outside the range of [0, 1).
+    """
+    if isinstance(value, float):
+        if not 0.0 < value <= 1.0:
```
[the remark this answers is not in the document]
Comment (RHammond2) [reply openoa/analysis/_analysis_validators.py:82]: Excellent catch, thank you!

**Comment by jordanperr** [comment]
Two thoughts:

**Thought 1:** How will this new paradigm extend to the workflow where analysis classes can "pull" data from an underlying datastore? For example, you might want to use MonteCarloAEP schema to pull all data necessary for a MonteCarloAEP from an ENTR warehouse. But then after pulling all the data, you might want some way to "exclude" one of the columns, say, to compare its impact on the AEP.

One possibility would be to extract the underlying PlantData object and then create two new MonteCarloAEP analyses using the same PlantData. Would that work?

I'm trying to think through if there might be a benefit to preserving the ability to just feed true/false into a flag in run().

**Thought 2:** Are PlantData and Analysis classes immutable? It didn't used to be. Please verify that no option in the run() function will mutate the PlantData and Analysis class. Or if they do, it should be noted somehow. The user will expect any combination of PlantData and Analysis class to be "run" multiple times with different options.

File: openoa/analysis/turbine_long_term_gross_energy.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (RHammond2) [reply openoa/analysis/turbine_long_term_gross_energy.py]: @ejsimley I was thinking on this a bit more, and I created and another tool in `_analysis_validators` to do this. Let me know what you think.

**Comment by RHammond2** [comment]
> Two thoughts:
> 
> **Thought 1:** How will this new paradigm extend to the workflow where analysis classes can "pull" data from an underlying datastore? For example, you might want to use MonteCarloAEP schema to pull all data necessary for a MonteCarloAEP from an ENTR warehouse. But then after pulling all the data, you might want some way to "exclude" one of the columns, say, to compare its impact on the AEP.
> 
> One possibility would be to extract the underlying PlantData object and then create two new MonteCarloAEP analyses using the same PlantData. Would that work?
> 
> I'm trying to think through if there might be a benefit to preserving the ability to include/exclude data in run().
> 
> **Thought 2:** Are PlantData and Analysis classes immutable? It didn't used to be. Please verify that no option in the run() function will mutate the PlantData and Analysis class. Or if they do, it should be noted somehow. The user will expect any combination of PlantData and Analysis class to be "run" multiple times with different options.

@jordanperr thanks for the comments! I updated the description of the PR, but just to put it all in one place:
- Only `analysis_type` gets updated in `PlantData` from an analysis class, but the `analysis.plant` object is also created as a deep copy of the original, so
- you can create many new analysis classes to compare results of various modifications with ease
- Naturally, if you modify an analysis parameter, the analysis class itself will be updated, otherwise you'd not be able to rerun an analysis with a new variation of a parameter, but what isn't modified is the underlying data.
- The only thing excluded from the `run()` parameters are those that modify which data are validated, and therefore changing the fundamental basis of the analysis, with the exception of `reanalysis_products`.
- For pulling analysis requirements from a data store, we still don't explicitly support this outside of grabbing requirements schemas for the analysis types in `openoa/schema/` or from `ANALYSIS_REQUIREMENTS`

Let me know if I've missed anything here, or if you still have any reservations.

**Pull request merged**